src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,50 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> tail;
+    private int size;
 
+    public void push(T value) {
+        if (tail == null) {
+            tail = new Node<>(value, null);
+        } else {
+            tail = new Node<>(value, tail);
+        }
```

Comment:
top is already null, we don't need first check
```suggestion
            tail = new Node<>(value, tail);
```